text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
+
+For example:
+
+```rust
+if an_option is Some(x) && x > 3 {
+    println!("{x}");
+}
+```
+
+The bindings in the pattern are not bound along any code path potentially
+reachable where the expression did not match:
+
+```rust
+if (an_option is Some(x) && x > 3) || (more_conditions /* x is not bound here*/) {
+    // x is not bound here
+} else {
+    // x is not bound here
+}
+// x is not bound here
+```
+
+The pattern may use alternation (within parentheses), but must have the same
+bindings in every alternative:
+
+```rust
+if color is (RGB(r, g, b) | RGBA(r, g, b, _)) && r == b && g < 10 {
+    println!("condition met")
+}
+
+// ERROR: `a` is not bound in all alternatives of the pattern
+if color is (RGB(r, g, b) | RGBA(r, g, b, a)) && r == b && g < 10 {
+    println!("condition met")
+}
+```
+
+`is` may appear anywhere a boolean expression is accepted:
+
+```rust
+func(x is Some(y) && y > 3);
+```
+
+The `is` operator may not appear as a statement; use `let` to bind a pattern in
+that context:
+
+```rust
+// ERROR: use `let` for this
+an_expression() is x;
+
+let x = an_expression();
+```
+
+# Reference-level explanation
+
+Add a new [operator
+expression](https://doc.rust-lang.org/reference/expressions/operator-expr.html),
+`IsExpression`:
+
+> **<sup>Syntax</sup>**\
+> _IsExpression_ :\
+> &nbsp;&nbsp; _Expression_ `is` _PatternNoTopAlt_
+
+Add `is` to the [operator
+precedence](https://doc.rust-lang.org/reference/expressions.html#expression-precedence)
+table, at the same precedence level as `==`, and likewise non-associative
+(requiring parentheses).
+
+Detect `is` appearing as a top-level statement and produce an error, with a
+rustfix suggestion to use `let` instead.
+
+# Drawbacks
+
+Introducing both the `is` operator and `let`-chains would provide two different
+ways to do a pattern match as part of a condition. Having more than one way to
+do something could lead people to wonder if there's a difference; we would need
+to clearly communicate that they serve similar purposes.
+
+An `is` operator will produce a name conflict with [the `is` method on
+`dyn Any`](https://doc.rust-lang.org/std/any/trait.Any.html#method.is) in the
+standard library, and with the (relatively few) methods named `is` in the
+ecosystem. This will not break any existing Rust code, as the operator will
+only exist in the Rust 2024 edition and newer. The Rust standard library and
+any other library that wants to avoid requiring the use of `r#is` in Rust 2024
+and newer could provide aliases of these methods under a new name; for
+instance, the standard library could additionally provide `Any::is` under a new
+name `is_type`.
+
+# Rationale and alternatives
```

Comment:
Something that I think is worth exploring here, even though I agree it's worse than the proposal, is the idea of just promoting let patterns to expressions. For example, allowing `f(let Some(y) = x && y > 5)`. This is consistent with let-chaining, but noticeably uncomfortable, and worth exploring as an example of further motivation for why `is` is the better option.

Reply:
I don't think it's even good for the language to promote patterns such as 
```rs
f(x is Some(y) && y > 5)
```
That promotes very obscure code which is really hard for people new or even intermediate to the language to even understand what is going on.

I'd much rather see patterns such as 

```rs
if x is Some(y) && y > 5 {
  f(true);
} else {
  f(false);
}
```
which while more verbose is less arcane. I agree that the first one looks prettier but there is a lot of information to unpack in one line, especially if you are new.

Reply:
Honestly, what you're describing to me is quite a stylistic choice and I don't think it's something that the language itself should have a say in, and maybe something that should be left in clippy lints.

What you've described to me is extremely similar to the common case of `if x { true } else { false }`, and it generally represents some failure to fully conceptualise booleans as data, rather than just conditions on branches. This is actually extremely common among programming in general and (IMO) represents a combination of failures in teaching, misconceptions accumulated from how other languages work, etc.

Like, to be clear, this isn't me saying you're wrong here-- it's a real problem and ignoring it is not a real solution. But in that regard, while failing to dig deep into why people prefer this more expanded version is ignoring it, it's also ignoring it to just say that the expanded version is better and not question it.

This is kind of why I think that the solution probably lies somewhere in clippy-- things such as `if x { true } else { false }` are warned in clippy lints, and similarly, your code would probably be changed back into mine after two passes, where the first notices that the `f(...)` could be factored outside the expression to f(if ... { true } else { false })` and then the second notices that you're doing `if x { true } else { false }`.